lib/parser/source/comment/associator.rb

```diff
@@ -80,74 +81,101 @@ def initialize(ast, comments)
       #
       def associate
         @mapping     = Hash.new { |h, k| h[k] = [] }
-        @comment_num = 0
+        @comment_num = -1
+        advance_comment
 
         advance_through_directives if @skip_directives
 
         process(nil, @ast)
 
-        @mapping
+        return @mapping
+      end
+
+      # #associate is broken for nodes which have the same content.
+      # e.g. 2 identical lines in the code will see their comments "merged".
+      # Using #associate_locations prevents this. The returned hash uses
+      # node.location as a key to retrieve the comments for this node.
+      def associate_locations
+        @map_using_node = false
```

Comment:
You also need to set `@map_using_node = true` on top of `associate`, otherwise calling `associate` after `associate_locations` will break the API contract.